src/box/sqldeserialize.c

```diff
@@ -0,0 +1,473 @@
+#include "diag.h"
+#include "execute.h"
+#include "lua/utils.h"
+#include "sqlInt.h"
+#include "sql/vdbe.h"
+#include "sql/vdbeInt.h"
+#include "execute.h"
+#include "schema.h"
+#include "session.h"
+#include "box.h"
+#include <small/ibuf.h>
+#include <msgpuck/msgpuck.h>
+
+#include "sqlparser.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+
+static struct Select *
+mp_decode_select(const char **data, bool subselect);
+
+static struct SrcList *
+mp_decode_select_from(const char **data);
+
+static struct ExprList *
+mp_decode_expr_list(const char **data);
+
+struct span_view {
+	const char * ptr;
+	uint32_t length;
+};
+
+// This is actually a bug in older gcc (i.e. gcc 4.6) and clang (i.e. 3.8)
+// which has been fixed in newer compiler versions (gcc 4.7+ and clang 6+)
+// ```
+//    struct span_view key = {0};
+//                            ^
+// sqldeserialize.c:410:29: error: missing field 'length' initializer [-Werror,-Wmissing-field-initializers]
+// This always was legal for C-style initializations, but well
+// who we are to fight with release_clang problems?
+#define SPAN_INIT() { .ptr = NULL, .length = 0}
+
+static inline char *
+sql_name_from_span(struct sql *db, const struct span_view *token)
+{
+	assert(token != NULL && token->ptr != NULL);
+	int sz = token->length;
+	char *ptr = sqlDbMallocRawNN(db, sz + 1);
+	if (ptr == NULL)
+		return NULL;
+
+	memcpy(ptr, token->ptr, sz);
+	ptr[sz] = '\0';
+
+	return ptr;
+}
+
+
+#define ON_(span, expected) \
+	if (span.length == (sizeof(expected) - 1) && \
+	    strncmp(expected, span.ptr, span.length) == 0)
+
+#define UNLESS_(span, expected) \
+	if (span.length != (sizeof(expected) - 1) || \
+	    strncmp(expected, span.ptr, span.length) != 0)
+
+#define EXPECT_MAP(data) ({ \
+		enum mp_type type = mp_typeof(**data); \
+		assert(type == MP_MAP); \
+		(void)type; \
```

Comment:
Can't you just write `assert(mp_typeof(**data) == MP_MAP);` instead of these three lines? After this, there will be three lines in the macro, and if this macro is inlined than there will be only two. I believe it will be more readable in this case.